lib/filter.js

```diff
@@ -0,0 +1,36 @@
+var utils = require('./utils');
+
+module.exports = Filter;
+
+function Filter() {
+  this.grep = /.*/;
+  this.invertGrep = false;
+  this.tags = [];
+  this.skipTags = [];
+}
+
+Filter.prototype.count = function(suite) {
+  var self = this;
+  var total = 0;
+  suite.eachTest(function(test){
+    if (self.shouldRun(test)) total++;
+  });
+  return total;
+};
+
+Filter.prototype.shouldRun = function(test) {
+  // --grep
+  var grepMatch = this.grep.test(test.fullTitle());
+  if (this.invertGrep) grepMatch = !grepMatch;
+  // --tags --skip-tags
+  var include = (!this.tags.length) || matchTags(test.ctx._tags, this.tags);
+  var exclude = this.skipTags.length && matchTags(test.ctx._tags, this.skipTags);
+  // final decision
+  return grepMatch && include && !exclude;
```

Comment:
Sharing so that someone points out my test is bad or wrong:

```js
const filterTags = ['ci']
const testTags = ['slow', 'vendor', 'thing', 'ci']

function matchTags (actualTags, against) {
  return against.some((tag) => {
    return actualTags.indexOf(tag) !== -1
  })
}

console.time('shouldRun')

let tests = 100 * 1000
while (--tests) matchTags(testTags, filterTags)

console.timeEnd('shouldRun')
```